### PR TITLE
Setup Eldev, Makefile, Tests and Update Github Actions

### PR DESCRIPTION
**Pull request by p4v4n**
- Setup Eldev and Makefile.
- Add 1 test with buttercup.
- Add github actions for tests and linting.
- Fix indentation and check-declare errors from Eldev.

### CODE REVIEW

**Comment by bbatsov** [comment]
You should also update the CI github action to use the Makefile.

**Comment by bbatsov** [comment]
While you're at this you can also add a `test` folder and one trivial buttercup test, so we'd have the unit test infrastructure in place.

**Comment by p4v4n** [comment]
> While you're at this you can also add a `test` folder and one trivial buttercup test, so we'd have the unit test infrastructure in place. (as in `clojure-mode`)

I have setup 1 unit test with buttercup and added running tests as part of the existing job in CI. Let me know if I update the job name or setup a separate job for tests.

File: .github/workflows/lint-emacs.yml [hunk 1]
```diff
@@ -29,3 +29,6 @@ jobs:
 
     - name: Lint the project
       run: eldev -dtT -C compile --warnings-as-errors
+
+    - name: Run tests
```
Comment (bbatsov) [inline .github/workflows/lint-emacs.yml:33]: If we keep the tests here we should probably rename the workflow to something like "CI".

**Comment by bbatsov** [comment]
> I have setup 1 unit test with buttercup and added running tests as part of the existing job in CI. Let me know if I update the job name or setup a separate job for tests.

I was just writing a comment about this. :D For me it's a bit better to have separate actions, as it makes it easier to see what worked in the output.

**Comment by p4v4n** [comment]
> > I have setup 1 unit test with buttercup and added running tests as part of the existing job in CI. Let me know if I update the job name or setup a separate job for tests.
> 
> I was just writing a comment about this. :D For me it's a bit better to have separate actions, as it makes it easier to see what ran/passed/failed in the output.

Cool. Will setup a new job for tests.

I have a couple of more questions.
1) What is currently running under lint is actually `eldev compile`. Should this be updated as well?
2) I am just getting familiar with github actions terminology.Should tests be a separate job inside this workflow/file or a separate workflow?
Adding tests as a separate workflow might be costly(assuming each workflow runs on a different VM) once we setup tests on every platform. If we are adding it to the same workflow/file the filename should probably be changed.

**Comment by bbatsov** [comment]
> What is currently running under lint is actually eldev compile. Should this be updated as well?

Yeah, I think it should just call `make lint`. 

> I am just getting familiar with github actions terminology.Should tests be a separate job inside this workflow/file or a separate workflow?

Separate job in the same workflow would be better.

**Comment by p4v4n** [comment]
> Separate job in the same workflow would be better.

I separated compile/lint/test into different jobs.

Current Status: lint is failing as `compat` package cannot be installed on CI. I have tried a few things but none of them fixed the issue so far.

**Comment by bbatsov** [comment]
I'm guessing @doublep might have some idea what's going wrong with `compat`.

**Comment by p4v4n** [comment]
> I'm guessing @doublep might have some idea what's going wrong with `compat`.

The compat issue was fixed after switching to emacs29.1 instead of snapshot.
Likely related to the below news. compat apparently is already part of emacs30 and will not be installed again. 
https://github.com/emacs-compat/compat/commit/c98e141d14114509caeda4e6752a1aad39a0cdb1

**Comment by bbatsov** [comment]
Yeah, that sounds like plausible explanation. I had missed this announcement. I guess now all that's left is to fixed the lint errors and squash all the related commits together.

**Comment by p4v4n** [comment]
> I guess now all that's left is to fixed the lint errors and squash all the related commits together.

Fixed the lint errors and cleaned up the commit history.

**Comment by bbatsov** [comment]
Great work! Thanks!

**Pull request merged**